src/main/java/org/beanpod/switchboard/controller/ChannelController.java

```diff
@@ -0,0 +1,120 @@
+package org.beanpod.switchboard.controller;
+
+import org.beanpod.switchboard.dao.ChannelDaoImpl;
+import org.beanpod.switchboard.dao.DecoderDaoImpl;
+import org.beanpod.switchboard.dao.EncoderDaoImpl;
+import org.beanpod.switchboard.dto.*;
+import org.beanpod.switchboard.dto.mapper.ChannelMapper;
+import org.beanpod.switchboard.dto.mapper.InputChannelMapper;
+import org.beanpod.switchboard.dto.mapper.OutputChannelMapper;
+import org.beanpod.switchboard.entity.ChannelEntity;
+import org.beanpod.switchboard.exceptions.ExceptionType;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import javax.transaction.Transactional;
+import java.util.List;
+import java.util.Optional;
+
+@RestController
+@RequestMapping("/channel")
+public class ChannelController {
+
+    static final String DELETE = " deleted";
+    @Autowired
+    ChannelDaoImpl channelService;
+    @Autowired
+    DecoderDaoImpl decoderService;
+    @Autowired
+    EncoderDaoImpl encoderService;
+
+    @Autowired
+    ChannelMapper channelMapper;
+    @Autowired
+    InputChannelMapper inputChannelMapper;
+    @Autowired
+    OutputChannelMapper outputChannelMapper;
+
+    @GetMapping
+    public List<ChannelDTO> retrieveAllChannels() {
+        List<ChannelEntity> channelEntities = channelService.getChannels();
+        return channelMapper.toChannelDTOs(channelEntities);
+    }
+
+    @GetMapping("/{id}")
+    public ResponseEntity<ChannelDTO> retrieveChannel(@PathVariable Long id) {
+        return channelService.findChannel(id).
+                map(ResponseEntity::ok).
+                orElseThrow(() -> new ExceptionType.DeviceNotFoundException(id.toString()));
+    }
+
+    @PostMapping
+    public ResponseEntity<ChannelDTO> createChannel(@RequestBody ChannelDTO channel) {
+        Optional<ChannelDTO> channelLookup = channelService.findChannel(channel.getId());
+        if (channelLookup.isPresent()) {
+            throw new ExceptionType.DeviceAlreadyExistsException(channel.getId().toString());
+        }
+        return ResponseEntity.ok(channelService.save(channel));
+    }
+
+    @PostMapping("/input/{id}/serial/{serial}")
```

Comment:
i chanced serial to decoder, not sure if you mean we should accept the model itself as a parameter(i.e. decoder object).
it would decrease the payload to just use their serial number

Reply:
Oh yeah, the way I phrased that was confusing. Keep it as is, `.../decoder/{serial}` to signify the serial number of the decoder of... I think you understood what I meant.